analyzer/src/diagnostic.rs

```diff
@@ -82,26 +82,49 @@ pub enum DiagnosticID {
 /// that are contained in a [Diagnostic] to emphasis/further explain the causes of the diagnostic.
 #[derive(Clone, PartialEq, Debug)]
 pub struct Observation {
+    /// Optional environment if the observation occurs on a different environment than the diagnostic's source
+    /// if none, this observation will target its diagnostic.
+    pub foreign_env: Option<SourceId>,
     /// Observed segment
     pub segment: SourceSegment,
     /// An optional help string to complete the observation
     pub help: Option<String>,
+    /// An optional tag to group observations
+    pub tag: Option<ObservationTag>,
 }
 
 impl Observation {
-    pub fn new(segment: SourceSegment) -> Self {
+    pub fn new(segment: impl Into<SourceSegment>) -> Self {
         Self {
-            segment,
+            segment: segment.into(),
+            foreign_env: None,
             help: None,
+            tag: None,
         }
     }
 
-    pub fn with_help(segment: SourceSegment, help: impl Into<String>) -> Self {
-        Self {
-            segment,
-            help: Some(help.into()),
-        }
+    pub fn within(mut self, env: SourceId) -> Self {
+        self.foreign_env = Some(env);
+        self
     }
+
+    pub fn with_help(mut self, help: impl Into<String>) -> Self {
+        self.help = Some(help.into());
+        self
+    }
+
+    pub fn with_tag(mut self, tag: ObservationTag) -> Self {
+        self.tag = Some(tag);
+        self
+    }
+}
+
+#[derive(Clone, PartialEq, Debug)]
+pub enum ObservationTag {
+    InFault,
+    Expected,
+    Declaration,
```

Comment:
`Expected` is for segments from which we expected something, `Declaration` is for segments with no particular expectation but got observed because they are relevant in the diagnostic (ex: `parameter is declared here`)